Makefile

```diff
@@ -44,14 +52,30 @@ export GOARCH ?= amd64
 
 
 # @todo allow user to run for a single $PKG only?
-PACKAGES := $(shell $(GO) list ./...)
+PACKAGES_CMD := $(GO) list ./...
 GO_EXCLUDE := /vendor/|.pb.go|.gen.go
-GO_FILES := $(shell find . -name '*.go' | grep -v -E '$(GO_EXCLUDE)')
+GO_FILES_CMD := find . -name '*.go' | grep -v -E '$(GO_EXCLUDE)'
 
 # Environment for tests, the directory containing istio and deps binaries.
 # Typically same as GOPATH/bin, so tests work seemlessly with IDEs.
 export ISTIO_BIN=${GO_TOP}/bin
 
+GO_VERSION_REQUIRED:=1.9
+
+# Parse out the x.y or x.y.z version and output a single value x*10000+y*100+z (e.g., 1.9 is 10900)
+# that allows the three components to be checked in a single comparison.
+VER_TO_INT:=awk '{split(substr($$0, match ($$0, /[0-9\.]+/)), a, "."); print a[1]*10000+a[2]*100+a[3]}'
+
+# using a sentinel file so this check is only performed once per version.  Performance is
+# being favored over the unlikely situation that go gets downgraded to an older version
+check-go-version: | ${ISTIO_BIN}/have_go_$(GO_VERSION_REQUIRED)
+${ISTIO_BIN}/have_go_$(GO_VERSION_REQUIRED):
+	@if test $(shell $(GO) version | $(VER_TO_INT) ) -lt \
+                 $(shell echo "$(GO_VERSION_REQUIRED)" | $(VER_TO_INT) ); \
+                 then printf "go version $(GO_VERSION_REQUIRED)+ required, found: "; $(GO) version; exit 1; fi
+	@mkdir -p ${ISTIO_BIN}
+	@touch ${ISTIO_BIN}/have_go_$(GO_VERSION_REQUIRED)
+
```

Comment:
Suggestion: we can have a simple python or perl script (python installed on most linux systems, mac as well) that simply runs the command and checks go version, and creates the placeholder file.

Also, isn't there a way in go to check the go compiler version and spit out an error if its below a certain version?

Reply:
Yeah, looking more at the other changes, you are using commands like find which are available only on osx, and linux/linux-emulated environments.. So, we could certainly stick a lot of these awkward parsing in shell script into a _simple_ python file (With zero dependencies).

Reply:
Are you thinking of Windows?  Things will fall apart on that OS, but most of the other parts of the build will already fail too even without this change.  There might be a different story with the Linux add-on for Windows 10, but otherwise there isn't going to be a python or even make there either (I think the closest that MSFT provides is nmake, and if someone installs cygwin then I'm not sure offhand what's available [in my experience its a bit of a kludge that people often give up on due to misc issues]).

Reply:
As for having go check the version, I wouldn't know about that offhand it it sounds nice if there's a way to do that.  This change came about because mixer's Makefile had a version check, but it didn't really work properly (it expected three version components and couldn't handle 2 components like 1.8), so I fixed it and Costin asked that I move it to the top-level Makefile hence this change.